pkg/ccl/testccl/sqlccl/sqlccl_test: TestReadCommittedReadTimestampNotSteppedOnCommit failed

pkg/ccl/testccl/sqlccl/sqlccl_test.TestReadCommittedReadTimestampNotSteppedOnCommit [failed](https://mesolite.cluster.engflow.com/invocations/default/a21dfbed-7151-44b7-90b6-94f1c6fcb16f?testReportRun=1&testReportShard=5&testReportAttempt=1#targets-Ly9wa2cvY2NsL3Rlc3RjY2wvc3FsY2NsOnNxbGNjbF90ZXN0) on master @ [42700f874211bdeaf18dd8c23bdcf18179fae456](https://github.com/cockroachdb/cockroach/commits/42700f874211bdeaf18dd8c23bdcf18179fae456):


```
=== RUN   TestReadCommittedReadTimestampNotSteppedOnCommit
    test_log_scope.go:172: test logs captured to: outputs.zip/logTestReadCommittedReadTimestampNotSteppedOnCommit1304633096
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:638: DRPC is enabled (via testserver factory defaults)
    read_committed_test.go:195: recording timestamp for {%!s(*kvpb.RequestUnion_Get=&{0xc012681260})}
    read_committed_test.go:195: recording timestamp for {%!s(*kvpb.RequestUnion_Get=&{0xc012681500})}
    read_committed_test.go:198: not recording timestamp for {%!s(*kvpb.RequestUnion_Put=&{0xc01265f440})}
    read_committed_test.go:195: recording timestamp for {%!s(*kvpb.RequestUnion_ConditionalPut=&{0xc0126da8c0})}
    read_committed_test.go:198: not recording timestamp for {%!s(*kvpb.RequestUnion_EndTxn=&{0xc0125d8b00})}
    read_committed_test.go:230: 
        	Error Trace:	pkg/ccl/testccl/sqlccl/read_committed_test.go:230
        	Error:      	"[1773853019.584154167,0 1773853019.589815071,0 1773853019.590397390,0]" should have 4 item(s), but has 3
        	Test:       	TestReadCommittedReadTimestampNotSteppedOnCommit
    panic.go:615: -- test log scope end --
test logs left over in: outputs.zip/logTestReadCommittedReadTimestampNotSteppedOnCommit1304633096
--- FAIL: TestReadCommittedReadTimestampNotSteppedOnCommit (0.54s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=1</code>
 - <code>shard=5</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReadCommittedReadTimestampNotSteppedOnCommit.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-61734

pkg/ccl/testccl/sqlccl/sqlccl_test.TestReadCommittedReadTimestampNotSteppedOnCommit [failed](https://mesolite.cluster.engflow.com/invocations/default/f88dc700-77ae-4594-8a5b-587b8912f56a?testReportRun=5&testReportShard=5&testReportAttempt=1#targets-Ly9wa2cvY2NsL3Rlc3RjY2wvc3FsY2NsOnNxbGNjbF90ZXN0) on master @ [ab9ecea4b27075b0c6d68604b403cd0de64d150b](https://github.com/cockroachdb/cockroach/commits/ab9ecea4b27075b0c6d68604b403cd0de64d150b):


```
=== RUN   TestReadCommittedReadTimestampNotSteppedOnCommit
    test_log_scope.go:172: test logs captured to: outputs.zip/logTestReadCommittedReadTimestampNotSteppedOnCommit3104956229
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:635: DRPC is enabled (via testserver factory defaults)
    read_committed_test.go:195: recording timestamp for {%!s(*kvpb.RequestUnion_Get=&{0xc0128e9740})}
    read_committed_test.go:195: recording timestamp for {%!s(*kvpb.RequestUnion_Get=&{0xc01297f200})}
    read_committed_test.go:198: not recording timestamp for {%!s(*kvpb.RequestUnion_Put=&{0xc0030a2510})}
    read_committed_test.go:195: recording timestamp for {%!s(*kvpb.RequestUnion_ConditionalPut=&{0xc001fcb540})}
    read_committed_test.go:198: not recording timestamp for {%!s(*kvpb.RequestUnion_EndTxn=&{0xc0128b7970})}
    read_committed_test.go:230: 
        	Error Trace:	pkg/ccl/testccl/sqlccl/read_committed_test.go:230
        	Error:      	"[1774415539.989619172,0 1774415539.999404284,0 1774415540.000036868,0]" should have 4 item(s), but has 3
        	Test:       	TestReadCommittedReadTimestampNotSteppedOnCommit
    panic.go:615: -- test log scope end --
test logs left over in: outputs.zip/logTestReadCommittedReadTimestampNotSteppedOnCommit3104956229
--- FAIL: TestReadCommittedReadTimestampNotSteppedOnCommit (0.57s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=5</code>
 - <code>shard=5</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReadCommittedReadTimestampNotSteppedOnCommit.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/ccl/testccl/sqlccl/read_committed_test.go
+++ b/pkg/ccl/testccl/sqlccl/read_committed_test.go
@@ -210,12 +210,26 @@ func TestReadCommittedReadTimestampNotSteppedOnCommit(t *testing.T) {
 	s, sqlDB, _ := serverutils.StartServer(t, params)
 	defer s.Stopper().Stop(ctx)
 
-	_, err := sqlDB.Exec(`CREATE TABLE kv (k TEXT, v INT) WITH (sql_stats_automatic_collection_enabled = false);`)
+	// Pin the write buffer size to the default (4MB). With a small metamorphic
+	// value, the buffer can overflow mid-transaction, causing a flush that
+	// prevents parallel commit and breaks the filter's EndTxn detection.
+	_, err := sqlDB.Exec(`SET CLUSTER SETTING kv.transaction.write_buffering.max_buffer_size = '4MB'`)
+	require.NoError(t, err)
+
+	_, err = sqlDB.Exec(`CREATE TABLE kv (k TEXT, v INT) WITH (sql_stats_automatic_collection_enabled = false);`)
+	require.NoError(t, err)
+
+	// Use a dedicated connection so we can collect a KV trace on failure.
+	conn, err := sqlDB.Conn(ctx)
+	require.NoError(t, err)
+	defer func() { require.NoError(t, conn.Close()) }()
+
+	_, err = conn.ExecContext(ctx, "SET TRACING = on, kv")
 	require.NoError(t, err)
 
 	// Create a read committed transaction that writes to three rows in three
 	// different statements and then commits.
-	tx, err := sqlDB.BeginTx(ctx, &gosql.TxOptions{Isolation: gosql.LevelReadCommitted})
+	tx, err := conn.BeginTx(ctx, &gosql.TxOptions{Isolation: gosql.LevelReadCommitted})
 	require.NoError(t, err)
 	_, err = tx.Exec(`INSERT INTO kv VALUES ('a', 1);`)
 	require.NoError(t, err)
@@ -224,6 +238,19 @@ func TestReadCommittedReadTimestampNotSteppedOnCommit(t *testing.T) {
 	_, err = tx.Exec(`INSERT INTO kv VALUES ('c', 3);`)
 	require.NoError(t, err)
 	require.NoError(t, tx.Commit())
+
+	_, err = conn.ExecContext(ctx, "SET TRACING = off")
+	require.NoError(t, err)
+
+	// Collect KV trace for debugging on failure.
+	traceRows := sqlutils.MakeSQLRunner(conn).QueryStr(t, "SHOW KV TRACE FOR SESSION")
+	defer func() {
+		if t.Failed() {
+			for _, row := range traceRows {
+				t.Logf("trace: %s", strings.Join(row, "\t"))
+			}
+		}
+	}()
 
 	// Verify that the transaction's read timestamp was not stepped on commit but
 	// was stepped between every other statement.

